### PR TITLE
Cache Webhook Authentication responses

### PR DESCRIPTION
**Pull request by cjcullen**
Add a simple LRU cache w/ 2 minute TTL to the webhook authenticator.

Kubectl is a little spammy, w/ >= 4 API requests per command. This also prevents a single unauthenticated user from being able to DOS the remote authenticator.

### CODE REVIEW

**Comment by cjcullen** [comment]
@erictune

**Comment by erictune** [comment]
@kubernetes/sig-auth

**Comment by erictune** [comment]
LGTM.  Maybe add a warning in pkg/apis/authentication.k8s.io/types.go that a cache may be used.

**Comment by liggitt** [comment]
I had envisioned a caching authorizer that wrapped an authorizer interface and just layered cache behavior on top. We did something similar in https://github.com/openshift/origin/blob/master/pkg/authorization/authorizer/cache/authorizer.go (though our authorizer interface is slightly different). 

I could see that being useful for many different authorizer impls, rather than making caching ad hoc per authz impl.

File: plugin/pkg/auth/authenticator/token/webhook/webhook.go [hunk 1]
```diff
@@ -36,30 +39,36 @@ var _ authenticator.Token = (*WebhookTokenAuthenticator)(nil)
...
-	}
-	result := w.RestClient.Post().Body(r).Do()
-	if err := result.Error(); err != nil {
-		return nil, false, err
+		Spec: v1beta1.TokenReviewSpec{Token: token},
 	}
-	if err := result.Into(r); err != nil {
-		return nil, false, err
+	if entry, ok := w.responseCache.Get(r.Spec); ok {
+		r.Status = entry.(v1beta1.TokenReviewStatus)
```
Comment (liggitt) [inline plugin/pkg/auth/authenticator/token/webhook/webhook.go]: Conditional cast? I'm still uneasy about dealing with specific serialized versions in this code, unlike the rest of the codebase

**Comment by cjcullen** [comment]
@liggit I agree. I think the webhook is the only authenticator that isn't making decisions purely in memory, so it's the only place where it really matters for right now.

Any thoughts on top-level cache on the union authenticator vs. cache for each? I think the former would be easier to be manage, but I may be overlooking something.

**Comment by cjcullen** [comment]
Looking into it a bit, a top-level cache for an authorizer on an `http.Request` probably wouldn't be very efficient. We'd probably need to group authenticators by type (token/basic/cert), or just wrap each individual one in its own cache. Since the other implementations are all in-memory anyway, I lean towards keeping this webhook-specific for now.

I think it'll be easier on the authorization side to do a cache around the union.

**Comment by liggitt** [comment]
We did a token authenticator cache wrapper at https://github.com/openshift/origin/blob/master/pkg/auth/authenticator/token/cache/cache.go

Probably segmented by type makes sense. 

I wouldn't block this on a cross-authn cache impl, but I'd at least consider what shape the external bits would take if that was available (I'd like to not change config flags if/when we make caching generally available)

**Comment by cjcullen** [comment]
If we add another cacheable token authenticator, we could reorganize the union authenticator, and add an `authentication-cache-ttl` flag. I think we'd be able to set up the options in apiserver to deprecate the webhook specific flag gracefully.

File: plugin/pkg/auth/authenticator/token/webhook/webhook.go (hunk 1 above)
Comment (cjcullen) [inline plugin/pkg/auth/authenticator/token/webhook/webhook.go]: The cache should only contain `TokenReviewStatus` objects. If we get something that isn't, we must be in a pretty bad state, and I don't know if there is a better answer than crashing.

**Comment by cjcullen** [comment]
@liggitt I see that you guys went with a 5m default for authn and authz. That's got me rethinking my 30 second default here. Is there a best practice on caching authentication decisions?

**Comment by deads2k** [comment]
> @liggitt I see that you guys went with a 5m default for authn and authz. That's got me rethinking my 30 second default here. Is there a best practice on caching authentication decisions?

Five minutes didn't exist for any particular reason.  In fact, the original pull has "5 minutes feels long. Make it tuneable" --deads2k (https://github.com/openshift/origin/pull/4873#discussion_r41295287).

I don't know of any customer that has ever changed it, but its a pretty easy setting so they may not have asked about it.

**Comment by cjcullen** [comment]
5 minutes actually doesn't feel that crazy for authentication. I think 5 minutes is a bit too long for authorization. Revoking/granting permissions (and expecting that to be enacted relatively quickly) is probably more common than revoking credentials, especially for a short lived access token.

**Comment by cjcullen** [comment]
I changed the default here to 2 minutes.

**Pull request merged**